pkg/operator/controller/cruft.go

```diff
@@ -186,6 +192,145 @@ func reconcileServiceAccounts(args *callbacks.ReconcileCallbackArgs) error {
+		}
+	}
+	return oldVersion, newVersion
+}
+
+func ensureNewVerIsStorage(args *callbacks.ReconcileCallbackArgs, realCrd, newCrd *extv1.CustomResourceDefinition) (*extv1.CustomResourceDefinition, error) {
+	realCrdCopy := realCrd.DeepCopy()
+	oldVersion, newVersion := getSpecVersions(realCrd)
+	if oldVersion != nil {
+		oldVersion.Storage = false
+	}
+	if newVersion == nil {
+		_, newVersion = getSpecVersions(newCrd)
+		realCrd.Spec.Versions = append(realCrd.Spec.Versions, *newVersion)
+	}
+	newVersion.Storage = true
+	if !reflect.DeepEqual(realCrdCopy, realCrd) {
+		err := args.Client.Update(context.TODO(), realCrd)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return realCrd, nil
+}
+
+func oldVerWasStoredVersion(realCrd *extv1.CustomResourceDefinition) bool {
+	for _, storedVersion := range realCrd.Status.StoredVersions {
+		if storedVersion == OldVersion {
+			return true
+		}
+	}
+	return false
+}
+
+func oldVerInCRD(currentCrd *extv1.CustomResourceDefinition) bool {
+	for _, currentVersion := range currentCrd.Spec.Versions {
+		if currentVersion.Name == OldVersion {
+			return true
+		}
+	}
+	return false
+}
+
+func rewriteOldObjects(args *callbacks.ReconcileCallbackArgs, crd *extv1.CustomResourceDefinition) error {
+	kind := crd.Spec.Names.Kind
+	gvk := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: OldVersion,
+		Kind:    kind,
+	}
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(gvk)
+	err := args.Client.List(context.TODO(), ul, &client.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, item := range ul.Items {
+		nn := client.ObjectKey{Namespace: item.GetNamespace(), Name: item.GetName()}
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(item.GetObjectKind().GroupVersionKind())
+		err = args.Client.Get(context.TODO(), nn, u)
+		if err != nil {
+			return err
+		}
+		err = args.Client.Update(context.TODO(), u)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func removeStoredVersion(args *callbacks.ReconcileCallbackArgs, crd *extv1.CustomResourceDefinition) error {
+	newStoredVersions := make([]string, 0)
+	for _, stored := range crd.Status.StoredVersions {
+		if stored != OldVersion {
+			newStoredVersions = append(newStoredVersions, stored)
+		}
+	}
+	crd.Status.StoredVersions = newStoredVersions
+	return args.Client.Status().Update(context.TODO(), crd)
+}
+func removeOldVersion(crd *extv1.CustomResourceDefinition) *extv1.CustomResourceDefinition {
+	newVersions := make([]extv1.CustomResourceDefinitionVersion, 0)
+	for _, version := range crd.Spec.Versions {
+		if version.Name != OldVersion {
+			newVersions = append(newVersions, version)
+		}
+	}
+	crd.Spec.Versions = newVersions
+	return crd
+}
+
+// Handle upgrade from clusters that had v1alpha1 as a storage version
+// and remove it from all CRDs managed by us
+func reconcileHandleOldVersion(args *callbacks.ReconcileCallbackArgs) error {
+	var err error
+
+	if args.State != callbacks.ReconcileStatePreUpdate {
```

Comment:
I would consider making this a "post read" callback that does different things based on the value of `args.currentObject` and `args.desiredObject`.

With a "post read" callback, currentObject is exactly the current object (has not been merged with desired yet)

1.  If there are stored versions not in desiredObject, add them to desired from current making sure they are not "stored" (desired has the current "stored" version)
2. If currentObject is missing any version in desired, exit (they will be added)
3. If stored versions is not a single value that is the most recent "desired" version, rewrite old objects and update CRD status stored version to be latest desired

Once the stored version is the latest version, this function should get called and do none of the above and that will lead to any versions not in desiredObject being deleted